cpp/src/arrow/testing/util.cc

```diff
@@ -170,8 +182,56 @@ Status MakeRandomByteBuffer(int64_t length, MemoryPool* pool,
   return Status::OK();
 }
 
-static int next_listen_port_ = 30001;
+int GetListenPort() {
+  // Get a new available port number by binding a socket to an ephemeral port
+  // and then closing it.  Since ephemeral port allocation tends to avoid
+  // reusing port numbers, this should give a different port number
+  // every time, even accross processes.
+  struct sockaddr_in sin;
+#ifdef _WIN32
+  SOCKET sock_fd;
+  auto sin_len = static_cast<int>(sizeof(sin));
+  auto errno_message = []() -> std::string {
+    return internal::WinErrorMessage(WSAGetLastError());
+  };
+#else
+#define INVALID_SOCKET -1
+#define SOCKET_ERROR -1
+  int sock_fd;
+  auto sin_len = static_cast<socklen_t>(sizeof(sin));
+  auto errno_message = []() -> std::string { return internal::ErrnoMessage(errno); };
+#endif
+
+#ifdef _WIN32
+  WSADATA wsa_data;
+  if (WSAStartup(0x0202, &wsa_data) != 0) {
+    ARROW_LOG(FATAL) << "Failed to initialize Windows Sockets";
+  }
+#endif
+
+  sock_fd = socket(AF_INET, SOCK_STREAM, IPPROTO_TCP);
+  if (sock_fd == INVALID_SOCKET) {
+    Status::IOError("Failed to create TCP socket: ", errno_message()).Abort();
+  }
+  // First bind to ('0.0.0.0', 0)
+  memset(&sin, 0, sizeof(sin));
+  sin.sin_family = AF_INET;
+  if (bind(sock_fd, reinterpret_cast<struct sockaddr*>(&sin), sin_len) == SOCKET_ERROR) {
+    Status::IOError("bind() failed: ", errno_message()).Abort();
```

Comment:
Could you close `sock_fd` before `Abort()`?

Reply:
I don't see the point, since we're aborting anyway. The fd is closed when the process exits.

Reply:
Just for good manner.
You're right. We don't need to close the fd.